sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -675,6 +676,22 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
       }
       if (newConditionOpt.isDefined) Join(left, right, joinType, newConditionOpt) else join
   }
+
+  /**
+   * Returns additional constraints which are not enforced on the result of join operations, but
+   * which can be enforced either on the left or the right side
```

Comment:
why not put this in `Join.validConstraints`? `LogicalPlan.constraints` should only contain constraints for the plab output, but `LogicalPlan.allConstraints` can contain more.

Reply:
I haven't put it there, because `constraints` is created from `allConstraints`, so adding them to `validConstraints` could have caused them to be part of `constraints` too.

Reply:
ah i see the problem. For left-anti join, although `Join.output` reuse the attributes from left child output, they are actually different attributes, e.g. Join may output null values, so we can't generate these constraints in `Join.validConstraints`.

I think we can override both `allConstraints` and `constraints`, to make sure these extra constraints appear in `allConstraints`, but not `constraints`.